File: Diagrammatize.Design.Pattern/code/cxx/Mediator/A1/ColleagueButton.cpp
#include "ColleagueButton.hpp"

ColleagueButton::ColleagueButton(const std::string& caption, QWidget *parent): QPushButton(caption.c_str(), parent) 
{}

ColleagueButton::~ColleagueButton() {}

void ColleagueButton::setMediator(Mediator* mediator_) {    // 保存Mediator
    mediator = mediator_;
}

void ColleagueButton::setColleagueEnabled(bool enabled) {   // Mediator下达启用/禁用的指示
    setEnabled(enabled);
}

File: Diagrammatize.Design.Pattern/code/python/Mediator/Sample/ColleagueButton.py
import tkinter as tk
from Colleague import Colleague

class ColleagueButton(tk.Button, Colleague):
    def __init__(self, parent, caption, command=None, font=None):
        self.mediator = None
        tk.Button.__init__(self, parent, text=caption, command=command, font=font)

    def setMediator(self, mediator):
        self.mediator = mediator

    def setColleagueEnabled(self, enabled):
        if enabled:
            self.config(state=tk.NORMAL)
        else:
            self.config(state=tk.DISABLED)

def main():
    import sys
    def quit():
        print("bye")
        sys.exit(1)

    root = tk.Tk()
    button = ColleagueButton(root, "hello", quit)
    button.pack()
    root.mainloop()

if __name__ == "__main__":
    main()
